pkg/test/framework/runtime/components/galley/client.go

```diff
@@ -0,0 +1,166 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package galley
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	"google.golang.org/grpc"
+	mcp "istio.io/api/mcp/v1alpha1"
+	mcpclient "istio.io/istio/pkg/mcp/client"
+	"istio.io/istio/pkg/mcp/testing/monitoring"
+	"istio.io/istio/pkg/test/framework/runtime/api"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+type client struct {
+	env     api.Environment
+	address string
+}
+
+func (c *client) waitForSnapshot(typeURL string, snapshot []map[string]interface{}) error {
+	conn, err := c.dialGrpc()
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+
+	urls := []string{typeURL}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	u := mcpclient.NewInMemoryUpdater()
+
+	cl := mcp.NewAggregatedMeshConfigServiceClient(conn)
+	mcpc := mcpclient.New(cl, urls, u, "", map[string]string{}, mcptestmon.NewInMemoryClientStatsContext())
+	go mcpc.Run(ctx)
+
+	_, err = retry.Do(func() (interface{}, bool, error) {
```

Comment:
I was thinking of the case where the client receives partial/incremental updates before the full desired state is pushed. I don't think the current retry logic handles that cases and would require additional hints, e.g. # of resources.

Reply:
The high-level snapshot model should work irrespective of whether it is incremental/partial. In other words, the model "here is a declarative set of things I expect as output in steady state" should work regardless.

The snapshot call establishes a new gRPC connection for every request, instead of reusing an existing gRPC connection. This probably is costly, but should be useful for avoiding any interference due to the stateful nature of the protocol.

The only failure case I can think of is a partial, in-between state being accepted as false positive. I don't think there is a good way to protect against this unless we add some mechanisms into Galley itself to reliably detect & track receipt of config changes. That is likely to be costly.

Reply:
Please disregard. I misread this earlier and didn't notice that the the got vs. want comparison itself is being retried.